processor/src/trace/virtual_bus/prover.rs

```diff
@@ -0,0 +1,103 @@
+/// This means, in a sense, that the opening proof for the final GKR evaluation is part of the STARK
+/// proof.
+///
+/// [1]: https://eprint.iacr.org/2023/1284
+pub struct VirtualBusProver<E, C, H>
+where
+    E: FieldElement<BaseField = Felt> + 'static,
+    C: RandomCoin<Hasher = H, BaseField = Felt>,
+    H: ElementHasher<BaseField = Felt>,
+{
+    composition_polynomials: Vec<Vec<Arc<dyn CompositionPolynomial<E>>>>,
+    _challenger: PhantomData<C>,
+}
+
+impl<E, C, H> VirtualBusProver<E, C, H>
+where
+    E: FieldElement<BaseField = Felt> + 'static,
+    C: RandomCoin<Hasher = H, BaseField = Felt>,
+    H: ElementHasher<BaseField = Felt>,
+{
+    /// Constructs a new [VirtualBusProver] given a set of random values for the GKR-LogUp relation.
+    ///
+    /// The constructor uses the randomness to do two things:
+    ///
+    /// 1. Compute the claimed value of the GKR-LogUp relation i.e., the generalization of the 0
+    /// constant in equation (3) in [1]. This constant can depend on the LogUp randomness as in
+    /// e.g., the overflow stack virtual table bus where such a constant represents entries in
+    /// the overflow stack which were present before the VM started executing.
+    ///
+    /// 2. The composition polynomials describing the numerators and denominators of the fractions
+    /// appearing on the left hand side of equation (3) in [1]. The denominators depend as well
+    /// on the LogUp randomness.
+    /// The composition polynomials are grouped into 4 sets that can be thought of as left/right
+    /// numerator/denominator. The reason for this can be understood by looking at the original
+    /// GKR protocol in [2] where the protocol therein uses the wiring predicates [\tilde{ADD}]
+    /// and [\tilde{MUL}] to encode the wiring of the circuit. For our purposes, we can do
+    /// away with wiring predicates using the left/right and numerator/denominator distinction.
+    ///
+    /// [1]: https://eprint.iacr.org/2023/1284
+    /// [2]: https://dl.acm.org/doi/10.1145/2699436
+    pub fn new(log_up_randomness: Vec<E>) -> Result<Self, Error> {
+        let (_claim, composition_polynomials) = generate(log_up_randomness)?;
+
+        Ok(Self {
+            composition_polynomials,
+            _challenger: PhantomData,
+        })
+    }
+
+    /// Returns the composition polynomials of the left/right numerators/denominators of
+    /// the GKR-LogUp relation.
+    fn composition_polynomials(&self) -> Vec<Vec<Arc<dyn CompositionPolynomial<E>>>> {
+        self.composition_polynomials.clone()
+    }
+
+    /// Proves the GKR-LogUp relation.
+    pub fn prove(&self, trace: &MainTrace, transcript: &mut C) -> GkrCircuitProof<E> {
+        // TODO: Optimize this so that we can work with base field element directly and thus save
+        // on memory usage.
+        let trace_len = trace.num_rows();
+        let mut mls: Vec<MultiLinearPoly<E>> = trace
+            .columns()
```

Comment:
Only the columns which figure in the virtual bus relation should be used here. At the moment all of the trace is used but this should changed in the furture